### PR TITLE
[stable9.1] Revert RHDS pagination PR 18294

### PR DESCRIPTION
**Pull request by PVince81**
Revert https://github.com/owncloud/core/pull/18294 to hopefully fix some performance issues.

@jvillafanez @butonic there was a conflict and I don't know enough about this to be able to judge how to solve it. I've left the conflict marker in the commit. Please have a look.

Maybe reverting is not necessary and only some parts of the original patch can be removed/disabled to increase performance.

### CODE REVIEW

**Comment by PVince81** [comment]
@butonic please review the code to make sure this can be used safely as a revert.

@jvillafanez please tell @davitol how to retest this to guarantee it doesn't break stuff

**Comment by butonic** [comment]
brought back nerver side sorting. otherwise we change behavior. The search code now looks exactly the same as before #18294. (https://github.com/owncloud/core/blob/2e8d8bf4efbd3e87a0459d10f921de6bd3848185/apps/user_ldap/lib/access.php#L974-L1059) minus some whitespace differences ...

**Comment by jvillafanez** [comment]
I removed that sorting due to the warning in the official documentation  (http://php.net/manual/en/function.ldap-sort.php).

It's fine if we keep this to check the performance but it should be removed eventually.

**Comment by butonic** [comment]
ok. ldap sort is deprecated with PHP7.0 We might need to rethink the sorting anyway if RHDS is broken. You can see weird sorting / inserting when just scrolling town in the users list.

I think it is ok to send this, for now.

**Comment by davitol** [comment]
Tetsed and no issues were found. LGTM :shipit: :+1:

File: apps/user_ldap/lib/Access.php [hunk 1]
```diff
@@ -1007,45 +1000,36 @@ private function search($filter, $base, $attr = null, $limit = null, $offset = n
...
 			}
+		}
+		$findings = array();
+		foreach($sr as $res) {
+			$findings = array_merge($findings, $this->ldap->getEntries($cr	, $res ));
+		}
 
-			$continue = $this->processPagedSearchStatus($sr, $filter, $base, $findings['count'],
-								$limit, $offset, $pagedSearchOK,
+		$this->processPagedSearchStatus($sr, $filter, $base, $findings['count'],
```
Comment (jvillafanez) [inline apps/user_ldap/lib/Access.php]: `findings['count']` needs to be reviewed. The array_merge could overwrite the count with another value.

File: apps/user_ldap/lib/Access.php [hunk 2]
```diff
@@ -1007,45 +1000,36 @@ private function search($filter, $base, $attr = null, $limit = null, $offset = n
...
+			//i.e. result do not need to be fetched, we just need the cookie
+			//thus pass 1 or any other value as $iFoundItems because it is not
+			//used
+			$this->processPagedSearchStatus($sr, $filter, $base, 1, $limit,
+											$offset, $pagedSearchOK,
+											$skipHandling);
+			return array();
+		}
+		// Do the server-side sorting
+		foreach(array_reverse($attr) as $sortAttr){
```
Comment (PVince81) [inline apps/user_ldap/lib/Access.php]: Can you change it ?

**Comment by jvillafanez** [comment]
Some data regarding the element sorting (with and without 4b5b926) based on timing the users page:

| without | with |
|-------|------|
| 15.22 | 11.67 |
| 15.27 | 10.9 |
| 15.88 | 10.97 |
| 10.7 | 10.71 |
| 10.35 | 10.49 |
| 11.04 | 10.92 |
| 12.19 | 10.71 |

Right now, merging results + sorting takes around 0.3 ms for 100 results and 0.1 ms for 50

File: apps/user_ldap/lib/Access.php [hunk 3]
```diff
@@ -1007,45 +1000,83 @@ private function search($filter, $base, $attr = null, $limit = null, $offset = n
...
+				}
+				// the rest of scenarios are ignored as we can't compare accordingly. return 0 just in case
+				return 0;
+			});
+
+			// restore the count at the end the array. The position shouldn't matter
+			$findings['count'] = $findingsCount;
+		}
+
+		$this->processPagedSearchStatus($sr, $filter, $base, $findings['count'],
```
Comment (jvillafanez) [inline apps/user_ldap/lib/Access.php]: We might need to investigate what happens with the `$findings['count']`. As said previously, the value can be overwritten with the array_merge, plus there might be a mismatch between `$findings['count']` and `count($findings)` (taking into account the extra "count" entry in the results). For now, I don't know what is the value we should be using.

**Comment by PVince81** [comment]
I really don't know what to do with this... a lot of effort went into this.

@jvillafanez is this still relevant for OC 10.0 ?

**Comment by jvillafanez** [comment]
Keep it around just in case.

**Comment by felixboehm** [comment]
Need context, not understanding this.

**Comment by butonic** [comment]
Considering this change for the ldap app in oc 10. there we no longer need the sort (for users) because we only fetch all users during sync. order has become irrelevant. let me check the group side.

**Comment by butonic** [comment]
In OC 10 the only end user visible part for group search is in the sharing dialogs. ldap_sort is being deprecated because it does client side sorting which might give random results when reaching the server side return-limit. server side sorting is planned. see https://github.com/smalot/ldap/issues/1#issuecomment-181904633 

we should just drop it.

**Comment by jvillafanez** [comment]
We should also ignore server-side sorting. We'll likely need proper indexing in the ldap to have acceptable performance, otherwise the ldap server will need to fetch all available results and sort them before sending them back to ownCloud, which would have a high cost for big installations. We also need to take into account the number of concurrent users hitting the ldap server.

**Comment by PVince81** [comment]
> we should just drop it.

happy to drop the PR if that's what you mean if OC 10 anyway obsoletes this

**Pull request closed without merging**